File: apps/agent-macos/Sources/EgressViewAgentCore/WorldAtlas.swift
import Foundation

/// Country outlines for the globe.
///
/// Reads the same TopoJSON the Web UI uses, so the two maps look alike, and
/// carries it inside the app: MapKit would fetch tiles from Apple, which an
/// agent expected to work without reaching anyone cannot depend on.
public struct WorldAtlas: Sendable {
    /// One closed ring in degrees, longitude first.
    public typealias Ring = [(longitude: Double, latitude: Double)]

    public struct Country: Sendable {
        public let code: String?
        public let name: String
        public let rings: [Ring]
    }

    public let countries: [Country]
    public var rings: [Ring] { countries.flatMap(\.rings) }

    public init(rings: [Ring]) {
        countries = [Country(code: nil, name: "", rings: rings)]
    }

    public static func bundled() throws -> WorldAtlas {
        guard let url = Bundle.module.url(
            forResource: "world-atlas-countries-110m", withExtension: "json"
        ) else {
            throw AtlasError.resourceMissing
        }
        return try WorldAtlas(topoJSON: Data(contentsOf: url))
    }

    public enum AtlasError: Error, Equatable {
        case resourceMissing
        case malformed
    }

    /// Decodes TopoJSON: arcs are delta-encoded integers that a shared
    /// transform turns back into degrees, and each shape refers to arcs by
    /// index rather than repeating shared borders.
    public init(topoJSON data: Data) throws {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let transform = root["transform"] as? [String: Any],
              let scale = transform["scale"] as? [Double], scale.count == 2,
              let translate = transform["translate"] as? [Double], translate.count == 2,
              let rawArcs = root["arcs"] as? [[[Double]]],
              let objects = root["objects"] as? [String: Any],
              let countryObject = objects["countries"] as? [String: Any],
              let geometries = countryObject["geometries"] as? [[String: Any]]
        else {
            throw AtlasError.malformed
        }

        let arcs: [Ring] = rawArcs.map { arc in
            var x = 0.0
            var y = 0.0
            return arc.compactMap { point in
                guard point.count >= 2 else { return nil }
                x += point[0]
                y += point[1]
                return (
                    longitude: x * scale[0] + translate[0],
                    latitude: y * scale[1] + translate[1]
                )
            }
        }

        func ring(for indices: [Int]) -> Ring {
            var result: Ring = []
            for index in indices {
                // A negative index means the arc is traversed backwards; the
                // encoding is ~i, so -1 is arc 0 reversed.
                let forward = index >= 0
                let position = forward ? index : ~index
                guard arcs.indices.contains(position) else { continue }
                let segment = forward ? arcs[position] : arcs[position].reversed()
                // The joining point is shared between consecutive arcs.
                result.append(contentsOf: result.isEmpty ? Array(segment) : Array(segment.dropFirst()))
            }
            return result
        }

        let english = Locale(identifier: "en_US")
        var codesByEnglishName: [String: String] = [:]
        for region in Locale.Region.isoRegions {
            guard let name = english.localizedString(forRegionCode: region.identifier) else { continue }
            if codesByEnglishName[name] == nil { codesByEnglishName[name] = region.identifier }
        }
        // Natural Earth's display names differ from CLDR for these regions.
        // Keep the exceptions explicit; guessing by a similar name could fill
        // the wrong country and would be worse than leaving one unfilled.
        let aliases: [String: String] = [
            "W. Sahara": "EH", "United States of America": "US",
            "Dem. Rep. Congo": "CD", "Dominican Rep.": "DO",
            "Falkland Is.": "FK", "Fr. S. Antarctic Lands": "TF",
            "Côte d'Ivoire": "CI", "Central African Rep.": "CF",
            "Congo": "CG", "Eq. Guinea": "GQ", "eSwatini": "SZ",
            "Palestine": "PS", "Myanmar": "MM", "Turkey": "TR",
            "Solomon Is.": "SB", "China": "CN", "Bosnia and Herz.": "BA",
            "Macedonia": "MK", "Trinidad and Tobago": "TT", "S. Sudan": "SS",
        ]

        var decodedCountries: [Country] = []
        for geometry in geometries {
            let properties = geometry["properties"] as? [String: Any]
            let name = properties?["name"] as? String ?? ""
            let code = aliases[name] ?? codesByEnglishName[name]
            var countryRings: [Ring] = []
            switch geometry["type"] as? String {
            case "Polygon":
                guard let polygon = geometry["arcs"] as? [[Int]] else { continue }
                countryRings.append(contentsOf: polygon.map(ring(for:)))
            case "MultiPolygon":
                guard let multi = geometry["arcs"] as? [[[Int]]] else { continue }
                for polygon in multi {
                    countryRings.append(contentsOf: polygon.map(ring(for:)))
                }
            default:
                continue
            }
            countryRings = countryRings.filter { $0.count >= 3 }
            if !countryRings.isEmpty {
                decodedCountries.append(Country(code: code, name: name, rings: countryRings))
            }
        }
        countries = decodedCountries
    }
}

File: apps/agent-windows/src/EgressView.Agent.Ui/WorldAtlas.cs
using System.Text.Json;
using System.Globalization;

namespace EgressView.Agent.Ui;

internal static class WorldAtlas
{
    internal sealed record Country(string? Code, string Name, IReadOnlyList<(double Lat, double Lon)[]> Rings);

    internal static IReadOnlyList<Country> Load()
    {
        try
        {
            var info = System.Windows.Application.GetResourceStream(
                new Uri("pack://application:,,,/EgressView.Agent.Ui;component/Resources/world-atlas-countries-110m.json", UriKind.Absolute));
            if (info is null) return [];
            using var document = JsonDocument.Parse(info.Stream);
            var root = document.RootElement;
            var transform = root.GetProperty("transform");
            var scale = transform.GetProperty("scale").EnumerateArray().Select(value => value.GetDouble()).ToArray();
            var translate = transform.GetProperty("translate").EnumerateArray().Select(value => value.GetDouble()).ToArray();
            var arcs = new List<(double Lat, double Lon)[]>();
            foreach (var rawArc in root.GetProperty("arcs").EnumerateArray())
            {
                var x = 0d; var y = 0d; var points = new List<(double, double)>();
                foreach (var rawPoint in rawArc.EnumerateArray())
                {
                    var values = rawPoint.EnumerateArray().Select(value => value.GetDouble()).ToArray();
                    x += values[0]; y += values[1];
                    points.Add((y * scale[1] + translate[1], x * scale[0] + translate[0]));
                }
                arcs.Add(points.ToArray());
            }

            var result = new List<Country>();
            var geometries = root.GetProperty("objects").GetProperty("countries").GetProperty("geometries");
            var codesByEnglishName = CountryCodesByEnglishName();
            var aliases = new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["W. Sahara"] = "EH", ["United States of America"] = "US",
                ["Dem. Rep. Congo"] = "CD", ["Dominican Rep."] = "DO",
                ["Falkland Is."] = "FK", ["Fr. S. Antarctic Lands"] = "TF",
                ["Côte d'Ivoire"] = "CI", ["Central African Rep."] = "CF",
                ["Congo"] = "CG", ["Eq. Guinea"] = "GQ", ["eSwatini"] = "SZ",
                ["Palestine"] = "PS", ["Myanmar"] = "MM", ["Turkey"] = "TR",
                ["Solomon Is."] = "SB", ["China"] = "CN", ["Bosnia and Herz."] = "BA",
                ["Macedonia"] = "MK", ["Trinidad and Tobago"] = "TT", ["S. Sudan"] = "SS",
            };
            foreach (var geometry in geometries.EnumerateArray())
            {
                var name = geometry.TryGetProperty("properties", out var properties) &&
                    properties.TryGetProperty("name", out var nameProperty)
                    ? nameProperty.GetString() ?? string.Empty : string.Empty;
                var code = aliases.GetValueOrDefault(name) ?? codesByEnglishName.GetValueOrDefault(name);
                var rings = new List<(double Lat, double Lon)[]>();
                var type = geometry.GetProperty("type").GetString();
                if (type == "Polygon") AddPolygon(geometry.GetProperty("arcs"), arcs, rings);
                else if (type == "MultiPolygon")
                    foreach (var polygon in geometry.GetProperty("arcs").EnumerateArray()) AddPolygon(polygon, arcs, rings);
                if (rings.Count > 0) result.Add(new Country(code, name, rings));
            }
            return result;
        }
        catch { return []; }
    }

    private static Dictionary<string, string> CountryCodesByEnglishName()
    {
        var result = new Dictionary<string, string>(StringComparer.Ordinal);
        foreach (var culture in CultureInfo.GetCultures(CultureTypes.SpecificCultures))
        {
            try
            {
                var region = new RegionInfo(culture.Name);
                result.TryAdd(region.EnglishName, region.TwoLetterISORegionName);
            }
            catch (ArgumentException) { }
        }
        return result;
    }

    private static void AddPolygon(JsonElement polygon, IReadOnlyList<(double Lat, double Lon)[]> arcs,
        ICollection<(double Lat, double Lon)[]> result)
    {
        foreach (var ringElement in polygon.EnumerateArray())
        {
            var ring = new List<(double Lat, double Lon)>();
            foreach (var indexElement in ringElement.EnumerateArray())
            {
                var encoded = indexElement.GetInt32();
                var index = encoded >= 0 ? encoded : ~encoded;
                if (index < 0 || index >= arcs.Count) continue;
                var segment = encoded >= 0 ? arcs[index] : arcs[index].Reverse().ToArray();
                ring.AddRange(ring.Count == 0 ? segment : segment.Skip(1));
            }
            if (ring.Count >= 3) result.Add(ring.ToArray());
        }
    }
}
